templates/unique-matrix-measure/main.go

```diff
@@ -0,0 +1,178 @@
+// package main holds the implementation of the routing template.
+package main
+
+import (
+	"log"
+	"time"
+
+	"github.com/nextmv-io/sdk/measure"
+	"github.com/nextmv-io/sdk/route"
+	"github.com/nextmv-io/sdk/run"
+	"github.com/nextmv-io/sdk/store"
+)
+
+func main() {
+	err := run.Run(solver)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+type stop struct {
+	ID        string `json:"id"`
+	Reference int    `json:"reference"`
+}
+
+// This struct describes the expected json input by the runner.
+// Features not needed can simply be deleted or commented out, but make
+// sure that the corresponding option in `solver` is also commented out.
+// In case you would like to support a different input format you can
+// change the struct as you see fit. You may need to change some code in
+// `solver` to use the new structure.
+type input struct {
+	Stops        []stop           `json:"stops"`
+	Vehicles     []string         `json:"vehicles"`
+	Starts       []stop           `json:"starts"`
+	Ends         []stop           `json:"ends"`
+	UniquePoints []route.Position `json:"unique_points"`
+	UniqueMatrix [][]float64      `json:"unique_matrix"`
+}
+
+// solver takes the input and solver options and constructs a routing solver.
+// All route features/options depend on the input format. Depending on your
+// goal you can add, delete or fix options or add more input validations. Please
+// see the [route package
+// documentation](https://pkg.go.dev/github.com/nextmv-io/sdk/route) for further
+// information on the options available to you.
+func solver(i input, opts store.Options) (store.Solver, error) {
+	// In case you directly expose the solver to untrusted, external input,
+	// it is advisable from a security point of view to add strong
+	// input validations before passing the data to the solver.
+
+	// Make sure we have a matrix of NxN where N is the number of unique points.
+	checkInput(i)
+
+	// Create a full matrix from the reduced unique matrix. The full matrix also
+	// has start/end locations for each vehicle incorporated. In this example we
+	// do not use any start/end locations.
+	// We go over each stop in each row of the full matrix and lookup it's stop
+	// references to access the cost from the reduced size matrix.
+	matrixSize := len(i.Stops) + 2*len(i.Vehicles)
+	fullMatrix := make([][]float64, matrixSize)
+
+	// Make all stops, starts and ends accessible from one slice.
+	joinedStops := joinStops(i, matrixSize)
+
+	for stopIndex1 := range fullMatrix {
+		// Create an row filled with 0.
+		fullMatrix[stopIndex1] = make([]float64, matrixSize)
+
+		// Get first reference index.
+		refIndex1 := joinedStops[stopIndex1].Reference
+		for stopIndex2 := range fullMatrix[stopIndex1] {
+			// Get second reference index and look up costs.
+			refIndex2 := joinedStops[stopIndex2].Reference
+			if refIndex1 != -1 && refIndex2 != -1 {
```

Comment:
If they are both `-1` then the cost in `fullMatrix` will be `0`. Is that what we want here?